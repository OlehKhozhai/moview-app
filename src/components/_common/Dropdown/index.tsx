import React from 'react';
import cn from 'classnames';

import DropdownOption from 'components/_common/DropdownOption';
import useClickOutside from 'hooks/useClickOutside';
import styles from './styles.module.scss';

type DropdownProps = {
  options: ReadonlyArray<string>;
  isOpen: boolean;
  value?: string;
  className?: string;
  onToggle: (e?: React.MouseEvent<HTMLElement>) => void;
  onClose: () => void;
  onOptionClick: (value: string) => void;
};

const Dropdown: React.FC<DropdownProps> = ({
  options,
  isOpen,
  value,
  className,
  onOptionClick,
  onToggle,
  onClose,
}) => {
  const dropdownRef = React.useRef(null);

  useClickOutside({ ref: dropdownRef, callback: onClose });

  return (
    <div className={cn(styles.root, className)} ref={dropdownRef}>
      <span className={styles[value ? 'value' : 'threeDots']} onClick={onToggle}>
        {value || '...'}
      </span>

      {isOpen && (
        <ul className={styles.dropdown}>
          {options.map((option) => {
            return <DropdownOption key={option} option={option} onOptionClick={onOptionClick} />;
          })}
        </ul>
      )}
    </div>
  );
};

export default React.memo(Dropdown);
