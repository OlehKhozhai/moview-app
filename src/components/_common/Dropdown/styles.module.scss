@import 'assets/styles/mixins/colors';

.root {
  position: relative;
}

.value {
  text-transform: uppercase;
  position: relative;
  cursor: pointer;

  &::after {
    content: '';
    position: absolute;
    width: 6px;
    height: 6px;
    border: solid $primary-border;
    border-width: 0 3px 3px 0;
    transform: rotate(45deg);
    right: -15px;
    top: 5px;
  }
}

.threeDots {
  transform: rotate(90deg);
  font-size: 30px;
  cursor: pointer;
  display: block;
}

.dropdown {
  position: absolute;
  cursor: pointer;
  top: 30px;
  left: -10px;
  width: 165px;
  background-color: $primary-light-bg;
  border-radius: 5px;
  border: 1px solid $tertiary-border;
  z-index: 1;
}

.dropdownOption {
  padding: 10px 15px;
  text-transform: capitalize;

  &:hover {
    background-color: $primary-dark-bg;
    border-radius: 5px;
  }
}
