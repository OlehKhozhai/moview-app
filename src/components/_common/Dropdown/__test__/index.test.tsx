import React from 'react';
import { create } from 'react-test-renderer';

import Dropdown from '../index';

describe('Dropdown component', () => {
  test('should render Dropdown component by default', () => {
    const component = create(
      <Dropdown
        isOpen={true}
        options={['test option 1', 'test option 2']}
        onOptionClick={() => undefined}
        onClose={() => undefined}
        onToggle={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Dropdown component with default value', () => {
    const component = create(
      <Dropdown
        value="Default value"
        isOpen={true}
        options={['test option 1', 'test option 2']}
        onOptionClick={() => undefined}
        onClose={() => undefined}
        onToggle={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Dropdown component with isOpen value of false ', () => {
    const component = create(
      <Dropdown
        isOpen={false}
        options={['test option 1', 'test option 2']}
        onOptionClick={() => undefined}
        onClose={() => undefined}
        onToggle={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should trigger onOptionClick, onClose and onToggle methods', () => {
    const handleOnOptionClick = jest.fn();
    const handleOnClose = jest.fn();
    const handleOnToggle = jest.fn();
    const component = create(
      <Dropdown
        isOpen={true}
        options={['test option 1', 'test option 2']}
        onOptionClick={handleOnOptionClick}
        onClose={handleOnClose}
        onToggle={handleOnToggle}
      />
    );

    component.root.props.onOptionClick();
    component.root.props.onClose();
    component.root.props.onToggle();

    expect(handleOnOptionClick).toHaveBeenCalled();
    expect(handleOnClose).toHaveBeenCalled();
    expect(handleOnToggle).toHaveBeenCalled();
  });
});
