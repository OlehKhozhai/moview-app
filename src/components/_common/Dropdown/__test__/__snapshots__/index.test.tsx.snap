// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should render Dropdown component by default 1`] = `
<div
  className="root"
>
  <span
    className="threeDots"
    onClick={[Function]}
  >
    ...
  </span>
  <ul
    className="dropdown"
  >
    <li
      className="root"
      onClick={[Function]}
    >
      test option 1
    </li>
    <li
      className="root"
      onClick={[Function]}
    >
      test option 2
    </li>
  </ul>
</div>
`;

exports[`Dropdown component should render Dropdown component with default value 1`] = `
<div
  className="root"
>
  <span
    className="value"
    onClick={[Function]}
  >
    Default value
  </span>
  <ul
    className="dropdown"
  >
    <li
      className="root"
      onClick={[Function]}
    >
      test option 1
    </li>
    <li
      className="root"
      onClick={[Function]}
    >
      test option 2
    </li>
  </ul>
</div>
`;

exports[`Dropdown component should render Dropdown component with isOpen value of false  1`] = `
<div
  className="root"
>
  <span
    className="threeDots"
    onClick={[Function]}
  >
    ...
  </span>
</div>
`;
